web/src/message_notifications.js

```diff
@@ -0,0 +1,447 @@
+        return true;
+    }
+
+    if (message.mentioned_me_directly) {
+        return true;
+    }
+
+    // The following blocks for 'wildcard mentions' and 'Followed topic wildcard mentions'
+    // should be placed below (as they are right now) the 'user_settings.enable_sounds'
+    // block because the global, stream-specific, and followed topic wildcard mention
+    // settings are wrappers around the personal-mention setting.
+    // wildcard mentions
+    if (
+        message.mentioned &&
+        stream_data.receives_notifications(message.stream_id, "wildcard_mentions_notify")
+    ) {
+        return true;
+    }
+
+    // Followed topic wildcard mentions
+    if (
+        message.mentioned &&
+        user_topics.is_topic_followed(message.stream_id, message.topic) &&
+        user_settings.enable_followed_topic_wildcard_mentions_notify
+    ) {
+        return true;
+    }
+
+    return false;
+}
+
+export function received_messages(messages) {
+    for (const message of messages) {
+        if (!message_is_notifiable(message)) {
+            continue;
+        }
+        if (!unread.message_unread(message)) {
+            // The message is already read; Zulip is currently in focus.
+            continue;
+        }
+
+        message.notification_sent = true;
+
+        if (should_send_desktop_notification(message)) {
+            process_notification({
+                message,
+                desktop_notify: notifications.granted_desktop_notifications_permission(),
+            });
+        }
+        if (should_send_audible_notification(message)) {
+            ui_util.play_audio($("#user-notification-sound-audio")[0]);
+        }
+    }
+}
+
+export function send_test_notification(content) {
+    received_messages([
+        {
+            id: Math.random(),
+            type: "test-notification",
+            sender_email: "notification-bot@zulip.com",
+            sender_full_name: "Notification Bot",
+            display_reply_to: "Notification Bot",
+            content,
+            unread: true,
+        },
+    ]);
+}
+
+function register_click_handlers({on_click_scroll_to_selected}) {
+    $("#compose_banners").on(
+        "click",
+        ".narrow_to_recipient .above_compose_banner_action_link",
+        (e) => {
+            const message_id = $(e.currentTarget).data("message-id");
+            narrow.by_topic(message_id, {trigger: "compose_notification"});
+            e.stopPropagation();
+            e.preventDefault();
+        },
+    );
+    $("#compose_banners").on(
+        "click",
+        ".sent_scroll_to_view .above_compose_banner_action_link",
+        (e) => {
+            const message_id = $(e.currentTarget).data("message-id");
+            message_lists.current.select_id(message_id);
+            on_click_scroll_to_selected();
+            compose_banner.clear_message_sent_banners(false);
+            e.stopPropagation();
+            e.preventDefault();
+        },
+    );
```

Comment:
Aren't these compose banners? I would think that they belong in the `compose_notifications.js` part of the split.

Reply:
Good catch. But that would reintroduce a `compose_notifications` → `narrow` → `message_edit` → `echo` → `compose_notifications` cycle. I guess we can avoid that with an `initialize` callback.